raster-foundry/blasterjs
# Add CLI and move to scripts directory

This PR does two main things:
  - moves the majority of scripts to a scripts directory instead of embedding them all into the `package.json`
 - adds a basic CLI to smooth the development experience

### Scripts directory
This PR added the following scripts in the `scripts` directory
 - bootstrap
 - build
 - clean
 - release
 - start
 - test

The remaining yarn based commands are:
 - build
 - start
 - test

which mostly reference the scripts within the scripts directory.


### CLI
This PR adds the `blaster` CLI. It should allow for more developers to contribute with less overhead, style policing, and chances for mistakes. It's pretty basic at this point so it can only:

- create a component in a specified package
- remove a component in a specified package
- create a constant in a specified package
- remove a constant in a specified package
- regenerate the package-level index files

Moving to a CLI setup also meant we needed to alter the way in which we were handling the package-level index file. This PR moves to three separate index files for each package.

**<package>/index.components.js** contains machine generated exports for all components within the package
**<package>/index.common.js** contains machine generated exports for all constants within the package
**<package>/index.js** is minimal file, pulling the component and constant index files in as well as providing the opportunity to do other exporting as well (like the `defaultTheme`) for example.

#### Commands
Each command has a single letter alias

**`blaster {g,generate} {component, constant} <packageName> <newThingName>`**
ex: `blaster g component core tooltip`, `blaster generate constant core direction`

Generates a component or constant in the specified package.
![image](https://user-images.githubusercontent.com/2442245/43852844-5e2eb6a8-9b0c-11e8-87b3-a333f4f73710.png)
![image](https://user-images.githubusercontent.com/2442245/43852868-703c3d34-9b0c-11e8-9225-c9dcaf13ef72.png)


**`blaster {d,destroy} {component, constant} <packageName> <thingToDestroyName>`**
ex: `blaster d component core tooltip`, `blaster destroy constant core direction`

Removes a component or constant in the specified package. Includes confirmation.

![image](https://user-images.githubusercontent.com/2442245/43852938-99f21e8c-9b0c-11e8-9e49-a638b6cbfc6d.png)
![image](https://user-images.githubusercontent.com/2442245/43852960-ac53b4a0-9b0c-11e8-853b-3a6a4cd688d1.png)


**`blaster {i,index}`**

Regenerates all three index files for each package as necesssary.

![image](https://user-images.githubusercontent.com/2442245/43852991-b91321a8-9b0c-11e8-98d6-1f4062dac1ee.png)


## Testing
- run each of the scripts listed above (don't go through the release, just see that it starts the process) and make sure they work as expected
- run `npm link` and make sure you can use the CLI (this is really important)
- try out the various CLI commands and make sure everything generates as you would expect

Once everything is confirmed as working I'll update the README.
